packages/admin/src/pages/scheme/[schemeId]/editors.getServerSideProps.tsx

```diff
@@ -0,0 +1,106 @@
+import { GetServerSidePropsContext } from 'next';
+import {
+  getSessionIdFromCookies,
+  getUserTokenFromCookies,
+} from '../../../utils/session';
+import {
+  getGrantScheme,
+  getSchemeEditors,
+  isSchemeOwner,
+} from '../../../services/SchemeService';
+
+export type EditorList = {
+  role: 'EDITOR' | 'OWNER';
+  email: string;
+  id: string;
+};
+
+export type Action = {
+  href: string;
+  label: string;
+  ariaLabel: string;
+};
+
+export type UnformattedEditorRow = {
+  key: string;
+  value: string;
+  action?: string | Action | JSX.Element;
+};
+
+const getEditorsServerSideProps = async ({
+  params,
+  req,
+}: GetServerSidePropsContext) => {
+  const { schemeId } = params as Record<string, string>;
+  const sessionCookie = getSessionIdFromCookies(req);
+  const userServiceJwt = getUserTokenFromCookies(req);
+
+  const isOwner = await isSchemeOwner(schemeId, sessionCookie);
+
+  const { name: schemeName } = await getGrantScheme(schemeId, sessionCookie);
+
+  if (!isOwner && req.url?.includes('manage-editors'))
+    return {
+      redirect: {
+        permanent: true,
+        destination: `/scheme/${schemeId}`,
+      },
+    };
+
+  function formatEditorRows({ email: key, role: value, id }: EditorList) {
+    const editorRow: UnformattedEditorRow = { key, value };
+    if (isOwner)
+      editorRow.action = {
+        href: `/scheme/${schemeId}/manage-editors/remove/${id}`,
+        label: 'Remove',
+        ariaLabel: `Remove ${key}`,
+      };
+    return editorRow;
+  }
+
+  const editorRows = (
+    await getSchemeEditors(schemeId, sessionCookie, userServiceJwt)
+  ).map(formatEditorRows);
+
+  editorRows.unshift({
+    key: 'Email',
+    value: 'Role',
+    action: isOwner ? 'Actions' : '',
+  } as UnformattedEditorRow);
```

Comment:
There is also _a lot_ of logic thats formatting the data for this table. We have the following functions:
- formatEditorRows
- formatEditorRow

and logic in between all that too:

```
  const editorRows = schemeEditors.map(formatEditorRows);

  editorRows.unshift({
    key: 'Email',
    value: 'Role',
    action: isOwner ? 'Actions' : '',
  });

  const mappedTableRows = editorRows.reduce(formatEditorRow, []);
```
I'm finding it very hard to follow what its actually doing and I suspect this can be cleaned up a lot

Reply:
The formatting logic here is just because the page provides data for both view and manage ui. In retrospect the api should probs handle this data formatting for us.